pkg/volume/util/device_util_linux.go

```diff
@@ -0,0 +1,86 @@
+// +build linux
+
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"errors"
+	"strings"
+)
+
+//FindMultipathDeviceForDevice given a device name like /dev/sdx, find the devicemapper parent
+func (handler *deviceHandler) FindMultipathDeviceForDevice(device string) string {
+	io := handler.get_io
+	disk, err := handler.FindDeviceForPath(device)
+	if err != nil {
+		return ""
+	}
+	sysPath := "/sys/block/"
+	if dirs, err := io.ReadDir(sysPath); err == nil {
+		for _, f := range dirs {
+			name := f.Name()
+			if strings.HasPrefix(name, "dm-") {
+				if _, err1 := io.Lstat(sysPath + name + "/slaves/" + disk); err1 == nil {
+					return "/dev/" + name
+				}
+			}
+		}
+	}
+	return ""
+}
+
+//FindDevicesForMultipathDevice given a disk name of /dev/dm-XX return a
+//slice including all the mapped devices
+//if the device isn't a mapped device return an empty slice
+func (handler *deviceHandler) FindDevicesForMultipathDevice(device string) []string {
+	io := handler.get_io
+	disk, err := handler.FindDeviceForPath(device)
+	if err != nil {
+		return make([]string, 0, 0)
+	}
+	sysPath := "/sys/block/" + disk + "/slaves/"
+	if dirs, err := io.ReadDir(sysPath); err == nil {
+		devs := make([]string, len(dirs), len(dirs))
+		for i, f := range dirs {
+			devs[i] = f.Name()
+		}
+		return devs
+	}
+	return make([]string, 0, 0)
+}
+
+//FindDeviceForVirtual Find the underlaying disk for a linked path such as /dev/disk/by-path/XXXX or /dev/mapper/XXXX
+// will return sdX or hdX etc, if /dev/sdX is passed in then sdX will be returned
+func (handler *deviceHandler) FindDeviceForPath(path string) (string, error) {
+	io := handler.get_io
+	parts := strings.Split(path, "/")
+	//if path /dev/hdX splits into "", "dev", "hdX" then we will
+	//return just the last part
+	if len(parts) == 3 && strings.HasPrefix(parts[1], "dev") {
+		return parts[2], nil
+	}
```

Comment:
Instead of repeating the same block of code before and after the `EvalSymlinks` you can just do it once after. `EvaSymLinks` will just return the original path if it is not a symlink. `EvaSymLinks("/dev/sda")`, for example, will return `"/dev/sda"`